sklearn/linear_model/ridge.py

```diff
@@ -141,7 +145,16 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
         # Only one penalty, we can solve multi-target problems in one time.
         K.flat[::n_samples + 1] += alpha[0]
 
-        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
+        try:
+            # Note: we must use overwrite_a=False in order to be able to
+            #       use the fall-back solution below in case a LinAlgError
+            #       is raised
+            dual_coef = linalg.solve(K, y, sym_pos=True,
+                                     overwrite_a=False)
+        except np.linalg.LinAlgError:
+            warnings.warn("Singular matrix in solving dual problem. Using "
+                          "least-squares solution instead.")
+            dual_coef = linalg.lstsq(K, y)[0]
```

Comment:
the normal equations should have an exact solution. the fact that
sympos=True means that it uses the optimized posv. however, least squares
gives the same solution, of course. i don't know which is faster, although
the latter calculates an svd which the former probably doesn't.

On Tuesday, January 13, 2015, Andreas Mueller <notifications@github.com>
wrote:

> In sklearn/linear_model/ridge.py
> <https://github.com/scikit-learn/scikit-learn/pull/3942#discussion-diff-22891602>
> :
>
> > @@ -141,7 +145,16 @@ def _solve_cholesky_kernel(K, y, alpha, sample_weight=None):
> >          # Only one penalty, we can solve multi-target problems in one time.
> >          K.flat[::n_samples + 1] += alpha[0]
> >
> > -        dual_coef = linalg.solve(K, y, sym_pos=True, overwrite_a=True)
> > +        try:
> > +            # Note: we must use overwrite_a=False in order to be able to
> > +            #       use the fall-back solution below in case a LinAlgError
> > +            #       is raised
> > +            dual_coef = linalg.solve(K, y, sym_pos=True,
> > +                                     overwrite_a=False)
> > +        except np.linalg.LinAlgError:
> > +            warnings.warn("Singular matrix in solving dual problem. Using "
> > +                          "least-squares solution instead.")
> > +            dual_coef = linalg.lstsq(K, y)[0]
>
> What is the reason to not use lstsq for all cases? Is it slower or does
> it provide a lower quality solution? I still didn't get that from your
> comment.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/3942/files#r22891602>.
>